Henrys work

birdday opened the pull request:
This is the somewhat massive set of codes used to (1) determine Henry's coefficients from adsorption data, (2) design arrays from the Henry's coefficients, and (3) predict sample compositions with the new algorithmic procedure. I tried to clean up the commits as best as I could, some are still a bit messy, or just large chunks of new code, but overall I'd say it's a lot cleaner than previous PRs, so hopefully it won't take us too long.

birdday committed:
Copied functions to avoid issues with relative addressing.

birdday committed:
Added functions/imports needed to compute analytical solution.

birdday committed:
Added import and basic functions needed for modified KLD work.

birdday committed:
Added tracking of array_pmfs_normalized for use with new kld method. Also removed return of element and array pmf for last cycle of algorithm since never used.

paulboone commented on sensor_array/analysis/execute_brute_force_analysis.py:
```diff
@@ -6,7 +6,7 @@
 import pandas as pd
 from datetime import datetime
-from mof_array.pmf.process_mass_data import (
+from brute_force_analysis import (
```
Changes need to be atomic. You can't change the module name without renaming the module.

paulboone commented on sensor_array/analysis/execute_brute_force_analysis.py:
```diff
@@ -27,21 +27,13 @@
 
 # Redefine key variables in yaml file
+sim_data = data['sim_data']
+exp_data = data['exp_data']
```
I'm glad this is no longer hard coded.

paulboone commented on sensor_array/analysis/execute_brute_force_analysis.py:
```diff
@@ -27,21 +27,13 @@
-
 # Import yaml file as dictoncary
-filepath = 'settings/process_config.sample.yaml'
+filepath = 'config_files/process_config.sample.yaml'
```
why are we loading from a .sample file?

paulboone commented on sensor_array/analysis/execute_genetic_algorithm_analysis.py:
```diff
@@ -9,32 +9,17 @@
-
 # Import yaml file as dictoncary
-filepath = '/Users/brian_day/Github-Repos/Sensor_Array/settings/process_config.yaml'
+filepath = 'config_files/process_config.sample.yaml'
```
why are we loading from a sample file?

paulboone commented on sensor_array/analysis/execute_genetic_algorithm_analysis.py:
```diff
@@ -9,32 +9,17 @@
 
 # Redefine key varaibles in yaml file
+sim_data = data['sim_data']
+exp_data = data['exp_data']
```
this method seems like a near duplicate of the above. Can they be combined? Can we have one settings load? Check out how I did my settings loading for the htsohm project: https://github.com/paulboone/htsohm/blob/pmaster/htsohm/config.py

paulboone commented on sensor_array/analysis/execute_genetic_algorithm_analysis.py:
```diff
@@ -4,7 +4,7 @@
 import pandas as pd
 from datetime import datetime
-from GeneticAlgorithm_BrianEdits import *
+from genetic_algorithm_analysis import *
```
again, changes must be atomic. You have to rename the directory at the same time.

paulboone commented on sensor_array/analysis/execute_brute_force_analysis.py:
```diff
@@ -13,16 +14,22 @@
     save_element_pmf_data,
     save_unbinned_array_pmf_data,
     save_binned_array_pmf_data,
     plot_binned_array_pmf_data)
```
this many imports is a sign you either should be using a namespace, i.e. calling these as brute_force_analysis.choose_arrays, or maybe the method that is calling all these should also be in the brute_force_analysis namespace

paulboone commented on sensor_array/analysis/brute_force_analysis.py:
```diff
@@ -320,7 +329,7 @@ def create_bins(gases, num_bins, mof_list, element_pmf_results):
         # lower_lim = lower_comp
         # upper_lim = upper_comp + (upper_comp-lower_comp)/(num_bins)
-        bin_points = np.linspace(lower_lim, upper_lim, num=num_bins+1, endpoint=True)
+        bin_points.append(np.linspace(lower_lim, upper_lim, num=num_bins+1, endpoint=True))
```
I need you to walk me through this commit ca6e013a5c11488af46f5a14a5dba656f585235b

paulboone replied:
Can you explain why we are using append here?

paulboone commented on sensor_array/analysis/brute_force_analysis.py:
```diff
@@ -641,6 +643,150 @@ def save_unbinned_array_pmf_data(gases, list_of_arrays, list_of_array_ids, all_a
+    for i in range(len(a)):
+        A_out = np.round(a[i]*100,3)
+        B_out = np.round(b[i]*100,3)
+        C_out = np.round(c[i]*100,3)
```
variables are not normally capitalized: https://www.python.org/dev/peps/pep-0008/#id45

paulboone commented on sensor_array/analysis/brute_force_analysis.py:
```diff
@@ -641,6 +643,150 @@ def save_unbinned_array_pmf_data(gases, list_of_arrays, list_of_array_ids, all_a
+    tax.right_axis_label("Oxygen", fontsize=axis_fontsize, offset = 0.15)
+    tax.bottom_axis_label("Nitrogen", fontsize=axis_fontsize, offset = 0.05)
+
+    plt.savefig('/Users/brian_day/Desktop/triplots/%s.png' % array_id, bbox_inches='tight')
```
shouldn't have hard code urls in the code base. usually, just saving them to the current directory is good.

paulboone commented on sensor_array/analysis/brute_force_analysis.py:
```diff
@@ -680,23 +826,33 @@ def plot_binned_array_pmf_data(gases, list_of_arrays, list_of_array_ids, bins, b
+
+    plt.legend([r'$CO_2$',r'$N_2$', r'$N_2$'], fontsize=8)
+    plt.tight_layout()
+    plt.savefig("/Users/brian_day/Desktop/binned_pmf_figs/%s.png" % (array))
```
hard coded path will break for everybody else

paulboone commented on sensor_array/helper/plot_GA_results_redo.py:
```diff
@@ -1,215 +0,0 @@
-import numpy as np
```
lets talk about this commit 9c54fbcd01fce9f182a5518f13d28c212bff1c1c

paulboone reviewed (changes requested):
let's talk this over after group meeting if you have time

birdday committed:
Temportary commit for PMF work - Rebasing up until this point.

birdday committed:
Array Design File.

birdday committed:
Sample Analysis Files.

birdday committed:
Henry's Coefficient Analysis from RASPA sims.

birdday committed:
Helper. - Changed File Permissions (outdated file).

birdday committed:
Helper. - Deleted unnecessary file.

birdday committed:
Helper. - Modified file for generating experimental data sets (both outdated files).

birdday committed:
Helper. - Cleaned up generate_compositions.py file.

birdday committed:
Added functions for henry's work to generate_compositions.py.'

birdday committed:
Sim. - Deleted unused file, write_gcmc_simulation.py.

birdday committed:
Sim. - Minor changes to RASPA template.

birdday committed:
Sim. - Changed handling of simulations to allow for multiple pressure…

…s and comps at same time.

birdday committed:
Sim. - Renamed write_compositions.py to write_simulations.py

birdday committed:
BFA. - Added plotting functions and auxillary functions for ternary plots.

birdday committed:
BFA. - Added options for assignemt of element probabilities.

birdday committed:
BFA. - Added functions needed to smooth mass data, and reintroduce random error into the mass data, to emulate error in RASPA.

birdday committed:
BFA. - Added file which examine the KLD for all experimental compositions of a single array.

birdday committed:
BFA. - Added functions needed to run one array at a time.

Also wrote new section of execute script which handles runnning one array at a time (to save RAM).

birdday committed:
BFA. - Fixes to handling of KLD and binning for all arrays.

birdday committed:
BFA. - Fixed name of import that should have been changed in last PR.

birdday committed:
BFA. - Imported function which reads results for use with one array at a time stuff.

birdday committed:
BFA. - Updated imports and remove some comments.

birdday committed:
GA. - Changed name of import which should have been part of last PR.

birdday committed:
GA. - Updated imports to use a config file.

birdday committed:
GA. - Updated calls in execte script (some still need to be part of config file).

birdday committed:
GA. - Added smoothing and error functions to GA.

birdday committed:
GA. - Added multiple error types for probability to GA code.

birdday committed:
Sim. - Added function to parse pressure file.

birdday committed:
Sim. - Joined unnecessarily separate files.

paulboone commented on sensor_array/analysis/brute_force_analysis.py:
```diff
@@ -191,9 +191,14 @@ def calculate_element_pmf(exp_results_full, sim_results_full, mof_list, stdev, m
+                a, b = 0, 2*float(max(all_masses_sim))
+                mu, sigma = float(mass_exp), float(stdev)
+                alpha, beta = ((a-mu)/sigma), ((b-mu)/sigma)
+            if type == 'percent':
```
use elif

paulboone commented on sensor_array/analysis/brute_force_analysis.py:
```diff
@@ -191,9 +191,14 @@ def calculate_element_pmf(exp_results_full, sim_results_full, mof_list, stdev, m
-            a, b = 0, float(max(all_masses_sim)) * (1 + mrange)
-            mu, sigma = float(mass_exp), float(stdev)*float(mass_exp)
-            alpha, beta = ((a-mu)/sigma), ((b-mu)/sigma)
+            if type == 'mass':
```
if there is a reference for these, might be helpful to have it here so somebody could verify where the math comes from.

paulboone commented on sensor_array/analysis/brute_force_analysis.py:
```diff
@@ -135,6 +136,140 @@ def import_simulated_data(sim_results_import, mof_list, mof_densities, gases):
     return sim_results_full
 
 
+def create_comp_list(simulated_data, mof_list, gases):
```
walk me through this over phone: 2e0ea26925572

paulboone commented on sensor_array/analysis/execute_brute_force_analysis_singlearray_allexp.py:
```diff
@@ -0,0 +1,101 @@
+    plot_element_mass_data,
+    plot_unbinned_array_pmf_data,
+    save_binned_array_pmf_data,
+    plot_binned_array_pmf_data)
```
this is screaming out for something like:

```
import brute_force_analysis as bfa
bfa.create_bins
# eetc
```

paulboone commented on sensor_array/analysis/brute_force_analysis.py:
```diff
@@ -391,6 +391,16 @@ def calculate_array_pmf(mof_array, element_pmf_results):
     return single_array_pmf_results
 
 
+def calculate_all_arrays_list(mof_list, num_mofs):
```
let's review this one over phone 8203b03d94a9280558401c7437fa573e870263ff

paulboone commented on sensor_array/analysis/execute_brute_force_analysis.py:
```diff
@@ -35,6 +35,7 @@
     plot_unbinned_array_pmf_data,
     save_binned_array_pmf_data,
     plot_binned_array_pmf_data)
+from genetic_algorithm_analysis import read_GA_results_messy
```
while this is technically an atomic commit, it doesn't functionally do anything, so this would typically be added elsewhere when this was needed

paulboone commented on sensor_array/analysis/genetic_algorithm_analysis.py:
```diff
@@ -193,6 +193,128 @@ def import_simulated_data(sim_results_import, mof_list, mof_densities, gases):
+    return comp_list, mole_fractions
+
+
+def moving_average_smooth(sim_results_import, mof_list, gases, comp_list, mole_fractions, num_points=1):
```
is this a duplicate of code elsewhere?

paulboone commented on sensor_array/analysis/genetic_algorithm_analysis.py:
```diff
@@ -350,43 +350,48 @@ def convert_experimental_data(exp_results_import, sim_results_import, mof_list,
-                    mof_temp_dict.update({ 'PMF_Range' : norm_probs_range[index] })
-                    # mof_temp_dict.update({ 'PMF_Exact' : norm_probs_exact[index] })
+                    mof_temp_dict['PMF_Range'] = norm_probs_range[index]
+                    mof_temp_dict['PMF_Exact'] = norm_probs_exact[index]
```
outputting the range and the exact info doesn't look like it is part of "Added multiple error types for probability to GA code." should either be a separate commit or the commit message should include something about this.

paulboone commented on sensor_array/analysis/execute_genetic_algorithm_analysis.py:
```diff
@@ -29,31 +29,42 @@
-# array_size_vect = [1,2,3,4,5]
-num_runs = 3
+array_size_vect = [5, 10, 15, 20, 25, 30, 35, 40, 45]
+num_runs = 1
```
commit message is vague and doesn't indicate these changes to the runs / array_size_vect

paulboone commented on sensor_array/helper/generate_compositions.py:
```diff
@@ -1,5 +1,341 @@
 
-gases = ['CO2','O2','N2']
+
+def create_uniform_comp_list(gases, gas_limits, spacing, filename=None, imply_final_gas_range=True, imply_final_gas_spacing=False, filter_for_1=True, round_at=None):
```
review this commit: 32585ab8

paulboone commented on sensor_array/analysis/array_analysis_SVD.py:
```diff
@@ -0,0 +1,291 @@
+The primary analysis method is the Singular Value Decomposition (SVD) mapping the composition space
+to the sensor arrya response space.
+"""
+
```
review this commit: 37a7868

paulboone reviewed (commented):
yes yes yes! This is much, much better. I still need to go through it with you, since there is a LOT of new code. Let's go over it over zoom.

birdday commented:
Okay great thanks! Sorry this is the codeI didn’t put much effort into rebasing because it wasn’t too important anymore, but gald you didn’t waste much time on it before.

I’m free all day today pretty much, and most of tomorrow, except for a short meeting with Xin around 11. Let me know when works for you.

From: Paul Boone <notifications@github.com>
Reply-To: WilmerLab/sensor_array_mof_adsorption <reply@reply.github.com>
Date: Thursday, July 16, 2020 at 10:55 AM
To: WilmerLab/sensor_array_mof_adsorption <sensor_array_mof_adsorption@noreply.github.com>
Cc: Brian Day <22bday@gmail.com>, Author <author@noreply.github.com>
Subject: Re: [WilmerLab/sensor_array_mof_adsorption] Henrys work (#19)


@paulboone commented on this pull request.

yes yes yes! This is much, much better. I still need to go through it with you, since there is a LOT of new code. Let's go over it over zoom.

________________________________

In sensor_array/analysis/brute_force_analysis.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455435857>:

> @@ -320,7 +329,7 @@ def create_bins(gases, num_bins, mof_list, element_pmf_results):

         # Jenna Bins

         # lower_lim = lower_comp

         # upper_lim = upper_comp + (upper_comp-lower_comp)/(num_bins)

-        bin_points = np.linspace(lower_lim, upper_lim, num=num_bins+1, endpoint=True)

+        bin_points.append(np.linspace(lower_lim, upper_lim, num=num_bins+1, endpoint=True))

Can you explain why we are using append here?

________________________________

In sensor_array/analysis/brute_force_analysis.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455436495>:

> @@ -191,9 +191,14 @@ def calculate_element_pmf(exp_results_full, sim_results_full, mof_list, stdev, m

         for mass_exp in all_masses_exp:

             probs_range = []

             probs_exact = []

-            a, b = 0, float(max(all_masses_sim)) * (1 + mrange)

-            mu, sigma = float(mass_exp), float(stdev)*float(mass_exp)

-            alpha, beta = ((a-mu)/sigma), ((b-mu)/sigma)

+            if type == 'mass':

+                a, b = 0, 2*float(max(all_masses_sim))

+                mu, sigma = float(mass_exp), float(stdev)

+                alpha, beta = ((a-mu)/sigma), ((b-mu)/sigma)

+            if type == 'percent':

use elif

________________________________

In sensor_array/analysis/brute_force_analysis.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455437258>:

> @@ -191,9 +191,14 @@ def calculate_element_pmf(exp_results_full, sim_results_full, mof_list, stdev, m

         for mass_exp in all_masses_exp:

             probs_range = []

             probs_exact = []

-            a, b = 0, float(max(all_masses_sim)) * (1 + mrange)

-            mu, sigma = float(mass_exp), float(stdev)*float(mass_exp)

-            alpha, beta = ((a-mu)/sigma), ((b-mu)/sigma)

+            if type == 'mass':

if there is a reference for these, might be helpful to have it here so somebody could verify where the math comes from.

________________________________

In sensor_array/analysis/brute_force_analysis.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455437671>:

> @@ -135,6 +136,140 @@ def import_simulated_data(sim_results_import, mof_list, mof_densities, gases):

     return sim_results_full





+def create_comp_list(simulated_data, mof_list, gases):

walk me through this over phone: 2e0ea26<https://github.com/WilmerLab/sensor_array_mof_adsorption/commit/2e0ea26925572b8af3681a10b58acbb5d4e79897>

________________________________

In sensor_array/analysis/execute_brute_force_analysis_singlearray_allexp.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455438194>:

> +    calculate_all_arrays_list,

+    calculate_all_arrays,

+    calculate_single_array_kld,

+    create_bins,

+    create_comp_set_dict,

+    bin_compositions,

+    bin_compositions_single_array,

+    calculate_kld,

+    choose_arrays,

+    assign_array_ids,

+    save_element_pmf_data,

+    save_unbinned_array_pmf_data,

+    plot_element_mass_data,

+    plot_unbinned_array_pmf_data,

+    save_binned_array_pmf_data,

+    plot_binned_array_pmf_data)

this is screaming out for something like:

import brute_force_analysis as bfa

bfa.create_bins

# eetc

________________________________

In sensor_array/analysis/brute_force_analysis.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455438654>:

> @@ -391,6 +391,16 @@ def calculate_array_pmf(mof_array, element_pmf_results):

     return single_array_pmf_results





+def calculate_all_arrays_list(mof_list, num_mofs):

let's review this one over phone 8203b03<https://github.com/WilmerLab/sensor_array_mof_adsorption/commit/8203b03d94a9280558401c7437fa573e870263ff>

________________________________

In sensor_array/analysis/execute_brute_force_analysis.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455438955>:

> @@ -35,6 +35,7 @@

     plot_unbinned_array_pmf_data,

     save_binned_array_pmf_data,

     plot_binned_array_pmf_data)

+from genetic_algorithm_analysis import read_GA_results_messy

while this is technically an atomic commit, it doesn't functionally do anything, so this would typically be added elsewhere when this was needed

________________________________

In sensor_array/analysis/genetic_algorithm_analysis.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455439646>:

> +                temp_dict[gas] = float(row[gas])

+            comp_list.extend([temp_dict])

+

+    # Extracta  a list of all possible mol fractions for each individual gas

+    mole_fractions = {}

+    for gas in gases:

+        temp_list = []

+        for row in comp_list:

+            temp_list.extend([float(row[gas])])

+        temp_list_no_dups = list(set(temp_list))

+        temp_list_no_dups.sort()

+        mole_fractions[gas] = temp_list_no_dups

+    return comp_list, mole_fractions

+

+

+def moving_average_smooth(sim_results_import, mof_list, gases, comp_list, mole_fractions, num_points=1):

is this a duplicate of code elsewhere?

________________________________

In sensor_array/analysis/genetic_algorithm_analysis.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455843797>:

>

             # Update dictionary with pmf for each MOF

             new_temp_dict = []

             # Initialize the dictioary

             if mof_temp_dict == []:

                 for index in range(len(norm_probs_range)):

                     mof_temp_dict = all_results_sim[index].copy()

-                    mof_temp_dict.update({ 'PMF_Range' : norm_probs_range[index] })

-                    # mof_temp_dict.update({ 'PMF_Exact' : norm_probs_exact[index] })

+                    mof_temp_dict['PMF_Range'] = norm_probs_range[index]

+                    mof_temp_dict['PMF_Exact'] = norm_probs_exact[index]

outputting the range and the exact info doesn't look like it is part of "Added multiple error types for probability to GA code." should either be a separate commit or the commit message should include something about this.

________________________________

In sensor_array/analysis/execute_genetic_algorithm_analysis.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455845382>:

>

 # --------------------------------------------------

 # ----- Run the Genetic Algorithm ------------------

 # --------------------------------------------------



-array_size_vect = [10, 15, 20, 25, 30, 35, 40, 45]

-# array_size_vect = [1,2,3,4,5]

-num_runs = 3

+array_size_vect = [5, 10, 15, 20, 25, 30, 35, 40, 45]

+num_runs = 1

commit message is vague and doesn't indicate these changes to the runs / array_size_vect

________________________________

In sensor_array/helper/generate_compositions.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455847902>:

>

-gases = ['CO2','O2','N2']

+

+def create_uniform_comp_list(gases, gas_limits, spacing, filename=None, imply_final_gas_range=True, imply_final_gas_spacing=False, filter_for_1=True, round_at=None):

review this commit: 32585ab<https://github.com/WilmerLab/sensor_array_mof_adsorption/commit/32585ab81bed7302a0b14ed7777eca0eddf9c43f>

________________________________

In sensor_array/analysis/array_analysis_SVD.py<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#discussion_r455848244>:

> @@ -0,0 +1,291 @@

+"""

+This will be used to design/analyze arrays where we leverage the Henry's Coefficients rather than the

+GCMC simulations of the full composition space (impossible scale up).

+

+This work closely follows the work presented by Sturluson et al. (Cory Simon's Group) in the paper

+'Curating Metal Organic-Frameworks to Compose Robust Gas Sensor Arrays in Dilute Conditions' (2019).

+

+The primary analysis method is the Singular Value Decomposition (SVD) mapping the composition space

+to the sensor arrya response space.

+"""

+

review this commit: 37a7868<https://github.com/WilmerLab/sensor_array_mof_adsorption/commit/37a78685111d13207239a144db63770832cb644b>

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/WilmerLab/sensor_array_mof_adsorption/pull/19#pullrequestreview-449424676>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AI7NJIH2SQGQC4R37NYMAALR34IGHANCNFSM4N7ZD2JA>.

paulboone commented on sensor_array/analysis/execute_brute_force_analysis_singlearray_allexp.py:
```diff
@@ -0,0 +1,101 @@
+# --------------------------------------------------
+array = ['ALUKIC']
+
+import numpy as np
```
import is on line 66?

paulboone commented on sensor_array/analysis/execute_brute_force_analysis_singlearray_allexp.py:
```diff
@@ -0,0 +1,101 @@
+        create_comp_set_dict(element_pmf_results, mof_list)
+
+    # Round Bins (Fix Later)
+    import numpy as np
```
also here

paulboone commented on sensor_array/helper/generate_compositions.py:
```diff
@@ -1,62 +1,5 @@
-            k = k+1
-print('\nTotal number of mixtures = ', k)
-comp_list = np.reshape(comp_list, (k,3))
-print('\nList of Compositions: \n', comp_list)
```
This commit is a mistake!

paulboone commented on sensor_array/helper/generate_compositions.py:
```diff
@@ -1,5 +1,341 @@
+#
+# breath_filepath = '/Users/brian_day/Desktop/HC_Work/breath-sample-results/'+type+'-samples/'
+# all_breath_samples, all_breath_samples_joined = load_breath_samples(breath_filepath, mof_list_filtered)
+# breath_filepath = '/Users/brian_day/Desktop/HC_Work/breath-sample-results/'+type+'-samples/'
```
commented crap should  be removed

paulboone commented on sensor_array/analysis/array_analysis_SVD.py:
```diff
@@ -0,0 +1,291 @@
+    MOF = re.sub(r'_eqeq', '', MOF)
+    MOF = re.sub(r'_EQeq', '', MOF)
+    MOF = re.sub(r'_EQEQ', '', MOF)
+    MOF = re.sub(r'_v2', '', MOF)
```
pep8 variables

paulboone commented on sensor_array/analysis/array_analysis_SVD.py:
```diff
@@ -0,0 +1,291 @@
+
+
+def clean_MOF_name(MOF):
+    MOF = re.sub(r'_eqeq', '', MOF)
```
shouldn't assign to an argument passed to a method, unless you intend the argument to be changed for caller.

paulboone commented on sensor_array/analysis/array_analysis_SVD.py:
```diff
@@ -0,0 +1,291 @@
+            temp_array.append(temp_row)
+            full_array.extend(temp_array)
+        return full_array
+
```
don't write your own file format!

paulboone commented on sensor_array/analysis/array_analysis_SVD.py:
```diff
@@ -0,0 +1,291 @@
+gases = ['acetone', 'ammonia', 'argon', 'CO2', 'hydrogen']
+gases = ['ammonia', 'argon', 'CO2']
+figure_path = '/Users/brian_day/Desktop/HC_Work/HenrysConstants_Analysis_Results/'
+
```
hard coded path

paulboone commented on sensor_array/analysis/execute_sample_analysis.py:
```diff
@@ -0,0 +1,239 @@
+num_cycles = data['num_cycles']
+added_error_value = data['added_error_value']
+seed_value = data['seed_value']
+results_filepath = data['results_filepath']
```
could just use data straight

paulboone commented on sensor_array/analysis/sample_analysis.py:
```diff
@@ -0,0 +1,1015 @@
+
+
+def check_prediciton_of_known_comp_range(henrys_data_no_air_effect, henrys_data_only_air_effect, henrys_data_combo, gases, mof_list, known_comp):
+
```
prediciton

paulboone commented on sensor_array/analysis/sample_analysis.py:
```diff
@@ -0,0 +1,1015 @@
+
+
+def check_prediciton_of_known_comp(henrys_data_no_air_effect, henrys_data_only_air_effect, henrys_data_combo, gases, mof_list, known_comp):
+
```
prediciton

paulboone commented on sensor_array/analysis/sample_analysis.py:
```diff
@@ -0,0 +1,1015 @@
+def plot_predicted_vs_true_for_all_breath_samples(gases, gas_limits, predicted_comps, true_comps, filepath=None, sort_data=False, sort_by=None):
+    from matplotlib.ticker import AutoMinorLocator
+    from matplotlib.ticker import FixedLocator
+
```
imports in the function? this implies the plots should be broken into a separate file

paulboone commented on sensor_array/analysis/execute_sample_analysis.py:
```diff
@@ -0,0 +1,239 @@
+import numpy as np
+import os
+
+import sample_analysis as sa
```
sample analysis is a possible refactor target to pair

paulboone commented on sensor_array/helper/execute_henrys_coefficient_analysis.py:
```diff
@@ -0,0 +1,64 @@
+# ----- Perform Analysis -----
```
bad commit, breaks the codebase

paulboone commented on sensor_array/helper/henrys_coefficient_analysis.py:
```diff
@@ -0,0 +1,718 @@
+    with open(filepath, 'r') as yaml_file:
+        data = yaml.load(yaml_file)
+    return data
+
```
if necessary and yaml doesn't have a way of doing this, then refactor into separate file for reuse

paulboone commented on sensor_array/analysis/sample_analysis.py:
```diff
@@ -1013,3 +1010,116 @@ def composition_prediction_algorithm_new(array, henrys_data, gases, comps, spaci
             comps = subdivide_grid_from_array(filtered_comps, gases, spacing)
+
+
+def create_uniform_comp_list(gases, gas_limits, spacing, filename=None, imply_final_gas_range=True, imply_final_gas_spacing=False, filter_for_1=True, round_at=None):
```
This would be a rejected commit because it fixes problems that are not "real" and could be fixed with how you are running the code.

paulboone commented on sensor_array/analysis/sample_analysis.py:
```diff
@@ -1123,3 +1124,29 @@ def read_kH_results(filename):
+        inv = np.linalg.pinv(array)
+
+    return inv
+
```
I don't like the name here because it is too generic and sounds already like np.linalg.inv

Can you just use pinv instead of this method?

paulboone reviewed (approved):
I approve merging this under duress, knowing that we need to refactor a lot of this codebase.
